### PR TITLE
Initial progress on improved projections wire-up

### PR DESCRIPTION
**Pull request by daniel-smith**
This PR introduces an improved approach for configuring projections. Mainly:

- Improved builder semantics
- Addition of a "bare bones" subscription, leaving state management to the consumer.
- ~~New DbContext projection context, which doesn't keep the DbContext alive for the lifetime of the process~~ UPDATE: this has been replaced with a more general "service" projection.
- Signaling an intial OnCatchingUp notification when subscribing to all events. This allows projection contexts to respond accordingly on the initial catching up phase. For example, to begin a transaction.
- OnCatchingUp is also signaled whenever the underlying subscription falls behind (if supported by the infrastructure - e.g. SQLStreamStore)
- Add an OnSaved hook, which allows consumers to add behaviour when events are saved. E.g. push a notification.

There is still more work to do in this space, which will be addressed in future PRs. For example:
- Support more than one projection
- ~~Proper options when configuring with `UseSqlStreamStore`~~ (this is now done)
- Error handling around subscription drops, or exceptions which might occur from any of the projection context hooks
- Complete/improve implementation for SQL projections
- Tests
- Cancellation tokens for event handlers

### CODE REVIEW

File: src/DomainBlocks.Projections/EventDispatcher.cs [hunk 1]
```diff
@@ -39,7 +42,7 @@ public sealed class EventDispatcher<TRawData, TEventBase>
     public async Task StartAsync(CancellationToken cancellationToken = default)
     {
         Log.LogDebug("Starting EventStream");
-        await ForAllContexts((c, ct) => c.OnSubscribing(ct), cancellationToken).ConfigureAwait(false);
+        await ForAllContexts((c, ct) => c.OnInitializing(ct), cancellationToken).ConfigureAwait(false);
```
Comment (carlosrfernandez) [inline src/DomainBlocks.Projections/EventDispatcher.cs:45]: Not for this PR but, the event dispatcher might need a StopAsync method. The CancellationToken supplied in the StartAsync will potentially never be signaled. 

Just making a note here

File: src/DomainBlocks.Projections.AspNetCore/EventDispatcherHostedService.cs [hunk 2]
```diff
@@ -43,6 +45,27 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
...
+    }
+
+    public async Task StartAsync(CancellationToken cancellationToken = default)
+    {
+        await _eventDispatcher.StartAsync(cancellationToken);
+    }
+
     public Task StopAsync(CancellationToken cancellationToken = default)
     {
         return Task.CompletedTask;
```
Comment (carlosrfernandez) [inline src/DomainBlocks.Projections.AspNetCore/EventDispatcherHostedService.cs:71]: Again, not for this PR. 
Also related to my other comment, we should think about properly stopping the event dispatcher and the subscriptions

File: src/DomainBlocks.Projections/EventDispatcher.cs (hunk 1 above)
Comment (daniel-smith) [reply src/DomainBlocks.Projections/EventDispatcher.cs:45]: Yep. I think we also want cancellation tokens for the event handlers, which we don't have at the moment

File: src/DomainBlocks.Projections.AspNetCore/EventDispatcherHostedService.cs (hunk 2 above)
Comment (daniel-smith) [reply src/DomainBlocks.Projections.AspNetCore/EventDispatcherHostedService.cs:71]: I think this should stop via the cancellation toke when the hosted service is disposed. But yes, a stop would be good. I think we might want to make our event dispatcher restartable for failure handling scenarios.

**Review by carlosrfernandez — APPROVED** [review]

File: src/Examples/Shopping.ReadModel/Startup.cs [hunk 3]
```diff
@@ -26,20 +41,73 @@ public Startup(IConfiguration configuration)
...
+        // **** New approach ****
+        services.AddHostedEventCatchUpSubscription((sp, subscriptionOptions) =>
+        {
+            subscriptionOptions.UseSqlStreamStore(o =>
             {
-                options.UseSqlStreamStorePublishedEvents()
-                    .UseEntityFramework<ShoppingCartDbContext, StreamMessageWrapper>()
-                    .WithProjections((builder, dbContext) =>
+                var connectionString = Configuration.GetValue<string>("SqlStreamStore:ConnectionString");
+                var settings = new PostgresStreamStoreSettings(connectionString);
```
Comment (carlosrfernandez) [inline src/Examples/Shopping.ReadModel/Startup.cs:65]: We might be introducing an issue here

Currently: we check for schemaName in the options. We should add this, as we use it

Comment (carlosrfernandez) [reply src/Examples/Shopping.ReadModel/Startup.cs:65]: Actually, nevermind this is on the consumer side. Ignore me

**Pull request closed without merging**